### PR TITLE
Parse BTF data if present and emit #line directives

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Parse BTF data if present and emit #line directives.

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: tools/bpf2c/btf_parser.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    std::string source_line;
+    uint32_t line_number;
+    uint32_t column_number;
+} line_info_t;
+
+typedef std::map<std::string, std::map<size_t, line_info_t>> btf_section_to_instruction_to_line_info_t;
+
+/**
+ * @brief Parse a .btf and .btf.ext section from an ELF file and Map of section
+ * name to map of offset to line information.
```
Comment (dthaler) [inline tools/bpf2c/btf_parser.h:22]: Can't parse grammar.
```suggestion
 * @brief Parse a .btf and .btf.ext section from an ELF file.
```

**Review by dthaler — COMMENTED** [review]
What I don't understand is how to make an ELF file on windows contain the sections this parses.
https://github.com/microsoft/ebpf-for-windows/blob/master/docs/tutorial.md for example says to run "clang -I ../../include -target bpf -Werror -O2 -c bpf.c -o bpf.o" but that doesn't add such a section.   The droppacket.o that is generated by the build does, though.  What is the difference?   I.e., what should tutorial.md say?

**Comment by Alan-Jowett** [comment]
> What I don't understand is how to make an ELF file on windows contain the sections this parses. https://github.com/microsoft/ebpf-for-windows/blob/master/docs/tutorial.md for example says to run "clang -I ../../include -target bpf -Werror -O2 -c bpf.c -o bpf.o" but that doesn't add such a section. The droppacket.o that is generated by the build does, though. What is the difference? I.e., what should tutorial.md say?

Add "-g" to the Clang command line.

**Comment by dthaler** [comment]
> > What I don't understand is how to make an ELF file on windows contain the sections this parses. https://github.com/microsoft/ebpf-for-windows/blob/master/docs/tutorial.md for example says to run "clang -I ../../include -target bpf -Werror -O2 -c bpf.c -o bpf.o" but that doesn't add such a section. The droppacket.o that is generated by the build does, though. What is the difference? I.e., what should tutorial.md say?
> 
> Add "-g" to the Clang command line.

Thanks!  Looks like libbpf and bpftool both have BTF parsing stuff in them too.   I will look and see if I can just use libbpf APIs to do everything I need or not.

**Comment by Alan-Jowett** [comment]
I propose replacing btf_parser with libbpf if it can get us the same information.

**Review by dthaler — APPROVED** [review]

**Comment by dthaler** [comment]
> I propose replacing btf_parser with libbpf if it can get us the same information.

I don't know yet if it can.  I'd recommend three stages:
1) What you have now, except move the btf parse line info to a static lib so I can use it from netsh.
2) Change the BTF defines to match those used by libbpf.  e.g. BTF_MAGIC for the magic number at the start of the section.
3) Use libbpf once we verify it can do it, or else contribute API additions to libbpf if not.   But don't block on this now since not blocking anything.

**Comment by Alan-Jowett** [comment]
@dthaler sounds good to me. 

Note:
As I just noted in #630 the best way to do this for netsh would be to gather this info in the verifier code as it is the one that pretty prints the BPF instruction stream.

**Comment by Alan-Jowett** [comment]
libbpf has the data:
https://github.com/libbpf/libbpf/blob/fec0813359a9dd68dff7edb9f221b0f18d94433d/src/libbpf_internal.h#L414

It just doesn't expose it via an API.

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**